improver/calibration/reliability_calibration.py

```diff
@@ -488,6 +488,178 @@ def process(self, cubes, coordinates=None):
+            (
+                old_bounds[0:lower],
+                np.array([[old_bounds[lower, 0], old_bounds[upper, 1]]]),
+                old_bounds[upper + 1 :],
+            )
+        )
+        new_points = np.mean(new_bounds, axis=1, dtype=np.float32)
+        new_bin_coord = iris.coords.DimCoord(
+            new_points, long_name="probability_bin", units=1, bounds=new_bounds
+        )
+        return new_bin_coord
+
+    def _combine_bin_pair(self, reliability_table):
+        """
+        Combine a pair of bins when non-monotonicity of the observation
+        frequency is detected. Iterate top-down from the highest forecast
+        probability bin to the lowest probability bin when combining the bins.
+        Only allow a single pair of bins to be combined.
+
+        Args:
+            reliability_table (iris.cube.Cube):
+                A reliability table to be manipulated.
+        """
+        observation_count = reliability_table.extract(
+            iris.Constraint(table_row_name="observation_count")
+        ).data
+        forecast_count = reliability_table.extract(
+            iris.Constraint(table_row_name="forecast_count")
+        ).data
+        forecast_probability_sum = reliability_table.extract(
+            iris.Constraint(table_row_name="sum_of_forecast_probabilities")
+        ).data
+        observation_frequency = np.array(observation_count / forecast_count)
+        for upper in np.arange(len(observation_frequency) - 1, 0, -1):
+            lower = upper - 1
+            (diff,) = np.diff(
+                [observation_frequency[lower], observation_frequency[upper]]
+            )
+            if diff < 0:
+                forecast_count = self._sum_pairs(forecast_count, lower, upper)
+                observation_count = self._sum_pairs(observation_count, lower, upper)
+                forecast_probability_sum = self._sum_pairs(
+                    forecast_probability_sum, lower, upper
+                )
+                new_bin_coord = self._create_new_bin_coord(
+                    reliability_table, lower, upper
+                )
+                reliability_table = reliability_table[..., 0:-1]
+                reliability_table.replace_coord(new_bin_coord)
+                obs_count_cube = reliability_table.extract(
+                    iris.Constraint(table_row_name="observation_count")
+                )
+                obs_count_cube.data = observation_count
+                forecast_count_cube = reliability_table.extract(
+                    iris.Constraint(table_row_name="forecast_count")
+                )
+                forecast_count_cube.data = forecast_count
+                forecast_prob_sum_cube = reliability_table.extract(
+                    iris.Constraint(table_row_name="sum_of_forecast_probabilities")
+                )
+                forecast_prob_sum_cube.data = forecast_probability_sum
+                reliability_table = iris.cube.CubeList(
```

Comment:
This whole extracting each cube and then merging back together is a messy way of doing this. I couldn't think of a better way to easily access the correct slice of the cube and insert the new data. I've had to do something similar in the function below. Any suggestions would be appreciated.